Refactor optional argument implementation for isotopes, snwgrain, therm1 and therm2

apcraig opened the pull request:
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Refactor isotope implementation with respect to optional arguments
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    CICE is bit-for-bit on cheyenne, full test suite.  https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#0bf0fdcabd00df95d1028a66f71784b0b68f1178.  Icepack regression testing bit-for-bit on cheyenne https://github.com/CICE-Consortium/Test-Results/wiki/icepack_by_hash_forks#0c573347b261eba1ab6a373273cc232a7feb1ee1
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on CICE or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/.)
    - [ ] Yes
    - [X] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [X] No 
- [X] Please provide any additional information or relevant details below:

- Use argcheck design
- Eliminate local copies of optional data
- Pass optional arguments down the calling tree
- Some minor format cleanup

apcraig commented:
This refactors the isotope implementation using the newer design, passing optional arguments down the calling tree and using the argcheck design.  If this seems reasonable, I will refactor other Icepack features in a similar way where reasonable.  Do we want a PR for each feature refactor or just one (or a few) PRs to cover multiple feature refactors?  Would like to get some feedback before moving forward.

This should be bit-for-bit and has no impact on the public interfaces.  Will run comprehensive tests to confirm TBD.

dabail10 reviewed (approved):
Sorry for not responding to the email. This looks fine to me.

eclare108213 commented on columnphysics/icepack_atmo.F90:
```diff
@@ -972,24 +941,17 @@ subroutine icepack_atm_boundary(sfctype,                   &
-                                   uvel=l_uvel, vvel=l_vvel,  &
-                                   Uref=l_Uref, zlvs=zlvs)
+                                   Qa_iso=Qa_iso,           &
+                                   Qref_iso=Qref_iso,       &
```
What happens if present(Qref_iso)=F?  Is this robust across all compilers/platforms?  You probably tested that previously...

apcraig replied:
This is robust.  The optional attribute and "present" state is passed down the calling tree.  As long as optional is set in the calling routine for this field if want to check it's "present" state (determined by the original caller) and/or the optional argument is not used unless it exists, all should be fine.  You can definitely get into trouble if you pass the optional argument down the calling tree and use it if it wasn't passed.  Hopefully, we'll avoid that with the logic we have setup.

eclare108213 commented on columnphysics/icepack_mechred.F90:
```diff
@@ -1825,33 +1826,41 @@ subroutine icepack_step_ridge (dt,           ndtd,         &
+         if (tr_iso) then
+            if (present(fiso_ocn)) then
+                ! OK
+            else
```
switch the logic to `if (.not.(present(tr_iso))) then` ?

apcraig replied:
Maybe.  But if you need 4 optional arguments, I think this syntax is clearer.  For example,

```
if (present(x1) .and. present(x2). and. present(x3).and.present(x4)) then
  !OK
else
  abort
endif
```

rather than

```
if (.not.present(x1) .or..not.present(x2).or..not.present(x3).or..not.present(x4)) then
  abort
endif
```

I prefer the former.  I know it's awkward too but I think it's cleaner.  I guess a question is whether

```
if (.not.(present(x1).and.present(x2).and.present(x3).and.present(x4))) then
  abort
endif
```

this would also work and maybe be best?  Happy for some input.

apcraig replied:
I ended up updating it using the final option above.

eclare108213 commented:
> Do we want a PR for each feature refactor or just one (or a few) PRs to cover multiple feature refactors?

One or a few to cover multiple features would be fine, but if something is particularly complicated, then just do that by itself.  E.g. I imagine that the thermodynamics will be complicated, since there are many options that feed into it.

apcraig commented:
I think it may be worth reviewing and merging this now.  Most of the therm1 and therm2 optional arguments have been refactored.  Still waiting on some additional insight about meltsliq, but that can be done in a separate PR.

apcraig commented:
Icepack and CICE test suites are bit-for-bit.

phil-blain commented on columnphysics/icepack_atmo.F90:
```diff
@@ -904,8 +896,28 @@ subroutine icepack_atm_boundary(sfctype,                   &
+
       character(len=*),parameter :: subname='(icepack_atm_boundary)'
 
+      !------------------------------------------------------------
+      ! Check optional arguments
+      ! This is used for ocean and ice sfctype, need to be careful
+      !------------------------------------------------------------
```
I'm not sure I understand this comment, why does it say 

> This is used for ocean and ice sfctype

but then we just check `sfctype == 'ice'` ? What is "This" referring to?

apcraig replied:
I need to clean up this comment.

phil-blain commented:
> This refactors the isotope implementation using the newer design, passing optional arguments down the calling tree and using the argcheck design. If this seems reasonable, I will refactor other Icepack features in a similar way where reasonable. 

I think this is a nice clean up, it improves readability,  I think. Also, gotta love a PR that removes more lines of code than it adds ;)

> Do we want a PR for each feature refactor or just one (or a few) PRs to cover multiple feature refactors? Would like to get some feedback before moving forward.

If I were doing it, I think I would probably do it that way:
- 1 PR per module (so 1 per file)
- In each PR, one commit per subroutine: removing `l_*` local variable, using `argcheck` to validate arguments, and adjusting calls to other subroutines if need be.

This would be very much more pleasant to review, at least for me :) 

Also I would avoid changing the order of the variables attributes (like moving `optional` to be the last attributes, with no other changes). This adds a lot of noise to the diffs. If we want to make such changes, that's OK, but I'd prefer if it were done in separate commits.

Thanks a lot for starting the work on this! :)

phil-blain commented on columnphysics/icepack_atmo.F90:
```diff
@@ -916,14 +896,28 @@ subroutine icepack_atm_boundary(sfctype,                   &
+      !------------------------------------------------------------
+
+      if (sfctype == 'ice') then
+      if (argcheck == 'always' .or. (argcheck == 'first' .and. first_call_ice)) then
```
The logic in this line is duplicated a few times. Would it make sense to isolate it to a function somewhere ? Then we could `use` that function instead of the `argcheck` variable.

The function could receive that boolean `fist_call` (or here `first_call_ice`) as argument and return a boolean...

eclare108213 commented on columnphysics/icepack_snow.F90:
```diff
@@ -1194,6 +1194,7 @@ subroutine drain_snow (nslyr, vsnon, aicen, &
          sliq = meltsliq  ! computed in thickness_changes
       endif
+!tcx??
       meltsliq = meltsliq
```
This line can be commented out.  This section of the code was rewritten a bit and corrected, and this line was the logical culmination of that process.

apcraig replied:
> > Do we want a PR for each feature refactor or just one (or a few) PRs to cover multiple feature refactors? Would like to get some feedback before moving forward.
> 
> If I were doing it, I think I would probably do it that way:
> 
>     * 1 PR per module (so 1 per file)
> 
>     * In each PR, one commit per subroutine: removing `l_*` local variable, using `argcheck` to validate arguments, and adjusting calls to other subroutines if need be.

I don't think 1 PR per module will work well.  The optional argument checking will depend on the underlying implementation and that needs to be coordinated.  Removing the l_ variables might result in further optimizations down the calling tree.  So maybe 1 PR per feature.  I also understand about general cleanup (formatting) with updating features.  I agree, it's just hard not to fix things when I see an opportunity, and I'm worried I may not circle back if I don't.  I'll try to be more careful.

eclare108213 commented on columnphysics/icepack_therm_itd.F90:
```diff
@@ -1852,15 +1853,15 @@ subroutine add_new_ice (ncat,      nilyr,      &
                enddo
             endif
 
-           frazil_conc = c0
```
This line is still present in a similar section of code, above.  The code is hard-wired for n_iso=3, apparently, so maybe this line should stay in case for some reason n_iso > 3?

eclare108213 commented on columnphysics/icepack_therm_vertical.F90:
```diff
@@ -2381,178 +2382,92 @@ subroutine icepack_step_therm1(dt, ncat, nilyr, nslyr,    &
+             (present(fswthru_vdf) .and. .not.present(fswthrun_vdf)) .or. &
+             (present(fswthru_idr) .and. .not.present(fswthrun_idr)) .or. &
+             (present(fswthru_idf) .and. .not.present(fswthrun_idf))) then
+            call icepack_warnings_add(subname//' error in fswthru [iv]d[rf] arguments')
```
@apcraig @dabail10  Is this the logic we want here?  Often it's fine to have category-aggregated variables and not keep the category-specific ones.

apcraig replied:
It's really the underlying implementation.  The implementation in Icepack with this update will always computed the aggregated variables if they are passed, and that requires the category specific arguments too.  We could change the implementation so both have to be passed for the category aggregate to be computed.  The optional argument checking is determined to a large degree on the underlying implementation.

eclare108213 reviewed (commented):
Just a few comments and questions -

apcraig commented:
> Just a few comments and questions -
> 
> I do worry a bit about Phil's comment [here](https://github.com/CICE-Consortium/Icepack/issues/255#issuecomment-508872338), regarding non-Fortran codes being able to use Icepack with Fortran keywords and optional arguments. What should we do about that?

One option is for the non-Fortran interfaces to NOT adopt the optional arguments in their interfaces/uses.  So non-Fortran might pass all arguments even if they are not needed.  I don't actually know whether that will work for all languages, but we should explore as an appropriate time.  I think for now, we should continue to add optional arguments moving forward.  It provide backwards compatibility in the current community.  If at some point in the future, we decide to not have optional arguments in icepack, we just remove the optional attribute on the Icepack interfaces and then update any interfaces in CICE/Icepack that do not conform.  Everyone in the community would have to do the same, also anytime Icepacks is updated with any new arguments.  That's the downside overall.  I don't think we should abandon optional arguments until there is a clear reason to do so.

apcraig commented on columnphysics/icepack_atmo.F90:
```diff
@@ -389,21 +377,13 @@ subroutine atmo_boundary_layer (sfctype,            &
+         do n = 1, n_iso
+            ratio = c0
+            if (Qa_iso(2) > puny) ratio = Qa_iso(n)/Qa_iso(2)
+            Qref_iso(n) = Qa_iso(n) - ratio*delq*fac
```
That's right, we have already checked the arguments in the highest level Icepack interface.  So they are safe to use here without checking.  That's our current design.  As I mentioned the other day, there are pros and cons to doing it this way.  A pro is that we check all the optional arguments early leveraging parameter flags and then we should be OK after that without having to check over and over.  A negative is that you have to keep track of what's going on in different layers of the code.

eclare108213 replied:
Okay, thank you.

eclare108213 commented:
> > Just a few comments and questions -
> > I do worry a bit about Phil's comment [here](https://github.com/CICE-Consortium/Icepack/issues/255#issuecomment-508872338), regarding non-Fortran codes being able to use Icepack with Fortran keywords and optional arguments. What should we do about that?
> 
> One option is for the non-Fortran interfaces to NOT adopt the optional arguments in their interfaces/uses. So non-Fortran might pass all arguments even if they are not needed. I don't actually know whether that will work for all languages, but we should explore as an appropriate time. I think for now, we should continue to add optional arguments moving forward. It provide backwards compatibility in the current community. If at some point in the future, we decide to not have optional arguments in icepack, we just remove the optional attribute on the Icepack interfaces and then update any interfaces in CICE/Icepack that do not conform. Everyone in the community would have to do the same, also anytime Icepacks is updated with any new arguments. That's the downside overall. I don't think we should abandon optional arguments until there is a clear reason to do so.

The voice of reason - sounds good to me.

eclare108213 commented:
Looks like there are just a few minor changes that could be made here.  @apcraig let us know when you're ready to merge.

apcraig commented:
cheyenne was down this week and i'm away from home, should be able to make progress next few days.  will update conservation when updates are ready.

apcraig commented:
I've updated the PR and running a full test suite now.  Added icepack_chkoptargflag logical function and cleaned up a few other things associated with comments above.  Feel free to review, will report updated results in the next day or so.

apcraig commented:
CICE testing is complete with latest changes and everything looks good.

phil-blain reviewed (approved):
Thanks, latest version looks good!
